### PR TITLE
Adding better support for Consul, enabling SSL/TLS client config

### PR DESCRIPTION
**Pull request by philipsoutham**
After having an issue with etcd after enabling SSL/TLS security features (turns out anything after v2.0.4 results in unnecessarily high cpu utilization when TLS communication between peers is enabled, but that's for another issue on another project), I decided to look at using consul as my back end. To my surprise, none of the "security" enablement flags (`-scheme`, `-client-ca-keys`, `-client-key`, `-client-cert`) were hooked up to the client config, further more the library/api that communicated with consul was depricated. This pull request updates the client library to one that is supported by hashicorp and allows the aforementioned flags to be passed down to the client handling the connection to consul.

### CODE REVIEW

File: backends/consul_client.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+	"github.com/hashicorp/consul/api"
+)
+
+// Client provides a wrapper around the consulkv client
+type ConsulClient struct {
+	client *api.KV
+}
+
+// NewConsulClient returns a new client to Consul for the given address
```
Comment (philipsoutham) [inline backends/consul_client.go]: I took the liberty of moving this to the `github.com/kelseyhightower/confd/backends` package as I was not able to use the great `Config` struct due to a circular import if it was in the `github.com/kelseyhightower/confd/backends/consul` packags. I'm open to making `NewConsulClient` "ugly" again enumerating all the values from the `Config` struct as the function arguments if this move is objectionable.

File: backends/consul_client.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+func NewConsulClient(config *Config) (*ConsulClient, error) {
```
Comment (philipsoutham) [inline backends/consul_client.go]: I took the liberty of moving this to the `github.com/kelseyhightower/confd/backends` package as I was not able to use the great `Config` struct due to a circular import if it was in the `github.com/kelseyhightower/confd/backends/consul` packags. I'm open to making `NewConsulClient` "ugly" again enumerating all the values from the `Config` struct as the function arguments if this move is objectionable.

File: config.go [hunk 3]
```diff
@@ -214,7 +214,7 @@ func setConfigFromFlag(f *flag.Flag) {
 		config.ClientCert = clientCert
 	case "client-key":
 		config.ClientKey = clientKey
-	case "client-cakeys":
+	case "client-ca-keys":
```
Comment (philipsoutham) [inline config.go]: Looks like this has been broken for a while.

**Comment by timfallmk** [comment]
:+1:

File: backends/consul_client.go (hunk 2 above)
Comment (bacongobbler) [inline backends/consul_client.go]: What's the circular import issue here? IMO the `backends` package is meant to be a [factory package](http://en.wikipedia.org/wiki/Factory_method_pattern) with each client implemented in its own `backends/<backend>` package, so we really should be following the same pattern:

```go
package consul

func New(config *Config) (*ConsulClient, error) {
...
}
```

File: backends/consul/client.go [hunk 4]
```diff
@@ -1,32 +1,61 @@
...
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
+	"net/http"
 	"path"
 	"strings"
 
-	"github.com/armon/consul-api"
+	"github.com/hashicorp/consul/api"
+	"github.com/kelseyhightower/confd/backends"
```
Comment (philipsoutham) [inline backends/consul/client.go]: @bacongobbler This is where the circular import occurs.

**Comment by philipsoutham** [comment]
Deliberately breaking pull request so I can show @bacongobbler the circular import I mentioned on in original PR.

File: backends/consul_client.go (hunk 2 above)
Comment (philipsoutham) [inline backends/consul_client.go]: Please see the deliberately broken update to the PR.

**Comment by bacongobbler** [comment]
Ah I see what you mean. None of the other clients take advantage of this config struct, so let's leave that to a future cleanup/refactor and just call it via `New(nodes []string, cert, key, caCert string) (*Client, error)` for now.

**Comment by philipsoutham** [comment]
K, requested changes made, PR is valid again.

**Comment by bacongobbler** [comment]
LGTM

@kelseyhightower, can you please review when you got time?

**Comment by bacongobbler** [comment]
@philipsoutham I just fixed the Godep issue in master, would you mind rebasing and squash your commits? i.e. https://github.com/philipsoutham/confd/commit/d4dbde3f621a18f6e65b874805fb94bec48f8544 doesn't need to show up in `git log` :)

**Comment by philipsoutham** [comment]
@bacongobbler History has been rewritten, the log should be clean now.

**Comment by kelseyhightower** [comment]
@philipsoutham @bacongobbler Rebase and we can merge.

**Comment by philipsoutham** [comment]
@kelseyhightower @bacongobbler PR has been rebased with master.

**Comment by kelseyhightower** [comment]
@philipsoutham Seems like a bunch of extra commits in here. I understand the Godep bits, but what are the other changes, how are they related?

**Comment by philipsoutham** [comment]
@kelseyhightower Don't know, I just rebased again which seems to have resolved the issue of other commits in the PR; should be clean agan.

**Comment by kelseyhightower** [comment]
@philipsoutham lgtm

**Pull request merged**